src/Stratis.Bitcoin.Features.BlockStore/Controllers/BlockStoreController.cs

```diff
@@ -149,29 +150,55 @@ public IActionResult GetBlock([FromQuery] SearchByHashRequest query)
+            }
+            catch (Exception e)
+            {
+                this.logger.LogError("Exception occurred: {0}", e.ToString());
+                return ErrorHelpers.BuildErrorResponse(HttpStatusCode.BadRequest, e.Message, e.ToString());
+            }
+        }
 
-                if (this.network.Consensus.IsProofOfStake)
-                {
-                    var posBlock = block as PosBlock;
 
-                    blockModel.PosBlockSignature = posBlock.BlockSignature.ToHex(this.network);
-                    blockModel.PosBlockTrust = new Target(chainedHeader.GetBlockTarget()).ToUInt256().ToString();
-                    blockModel.PosChainTrust = chainedHeader.ChainWork.ToString(); // this should be similar to ChainWork
+        /// <summary>
+        /// Retrieves the blocks from a given height onwards.
+        /// </summary>
+        /// <param name="query">An object containing the necessary parameters to search for a block.</param>
+        /// <returns><see cref="BlockModel"/> if block is found, <see cref="NotFoundObjectResult"/> if not found. Returns <see cref="IActionResult"/> with error information if exception thrown.</returns>
+        /// <response code="200">Returns data about the block or block not found message</response>
+        /// <response code="400">Block hash invalid, or an unexpected exception occurred</response>
+        [Route(BlockStoreRouteEndPoint.GetBlocks)]
+        [HttpGet]
+        [ProducesResponseType((int)HttpStatusCode.OK)]
+        [ProducesResponseType((int)HttpStatusCode.BadRequest)]
+        [ProducesResponseType((int)HttpStatusCode.NotFound)]
+        public IActionResult GetBlocks([FromQuery] SearchByHeightRequest query)
+        {
+            if (!this.ModelState.IsValid)
+                return ModelStateErrors.BuildErrorResponse(this.ModelState);
+
+            try
+            {
+                if (query.Height > this.chainIndexer.Tip.Height)
+                    return this.NotFound("No blocks found");
 
-                    if (this.stakeChain != null)
-                    {
-                        BlockStake blockStake = this.stakeChain.Get(blockId);
+                int tip = query.Height + query.NumberOfBlocks - 1;
+                if (tip > this.chainIndexer.Tip.Height)
+                {
+                    query.NumberOfBlocks -= (this.chainIndexer.Tip.Height - tip);
+                    tip = this.chainIndexer.Tip.Height;
+                }
 
-                        blockModel.PosModifierv2 = blockStake?.StakeModifierV2.ToString();
-                        blockModel.PosFlags = blockStake?.Flags == BlockFlag.BLOCK_PROOF_OF_STAKE ? "proof-of-stake" : "proof-of-work";
-                        blockModel.PosHashProof = blockStake?.HashProof?.ToString();
-                    }
+                List<ChainedHeader> chainedHeaders = this.chainIndexer.Tip.EnumerateToGenesis().Take(query.NumberOfBlocks).Reverse().ToList();
```

Comment:
Not entirely sure I understand this line - let's say for example the chain is at height 100, and we request 10 blocks at height 50. Intuitively I would be expecting blocks 50-59 to be returned back. This seems like it would retrieve the chained headers for the upper 10 blocks on the chain? Or am I misunderstanding?

Reply:
Good find. Fixed.